fileversion.go

```diff
@@ -218,3 +240,57 @@ func (self *FileVersion) GetFixedFileInfo() (*FixedFileInfo, error) {
 		FileSubtype: VerFileSubtype(ffi.FileSubtype),
 	}, nil
 }
+
+func (self *FileVersion) GetFileInfoTranslations() ([]FileInfoTranslation, error) {
+	var fit *FileInfoTranslation
+	var len uint32
+	err := wrappers.VerQueryValue(
+		&self.data[0],
+		syscall.StringToUTF16Ptr(`\VarFileInfo\Translation`),
+		(**byte)(unsafe.Pointer(&fit)),
+		&len)
+	if err != nil {
+		return nil, NewWindowsError("VerQueryValue", err)
+	}
+
+	result := make([]FileInfoTranslation, 0)
+	if len == 0 {
+		return result, nil
+	}
+
+	ti := fit
+	for i := 0; i < int(len/uint32(unsafe.Sizeof(*ti))); i++ {
+		result = append(result, FileInfoTranslation{Language: ti.Language, CodePage: ti.CodePage})
+		ti = (*FileInfoTranslation)(unsafe.Pointer(uintptr(unsafe.Pointer(ti)) + unsafe.Sizeof(*ti)))
+	}
+	return result, nil
+}
+
+func (self *FileVersion) GetStringFileInfo(translation FileInfoTranslation, stringName string) (string, error) {
+	var offset *uint16
+	var len uint32
+	err := wrappers.VerQueryValue(
+		&self.data[0],
+		syscall.StringToUTF16Ptr(fmt.Sprintf("\\StringFileInfo\\%04x%04x\\%s", translation.Language, translation.CodePage, stringName)),
+		(**byte)(unsafe.Pointer(&offset)),
+		&len)
+	if err != nil {
+		return "", NewWindowsError("VerQueryValue", err)
+	}
+	if len == 0 {
+		return "", nil
+	}
+	return LpstrToString(offset), nil
+}
+
+func (self *FileVersion) GetFirstStringFileInfo(stringName string) (string, error) {
```

Comment:
So, the approach taken here seems like it has the potential to be a bit inefficient.  If you repeatedly call `GetFirstStringFileInfo` to retrieve multiple strings, it will call `GetFileInfoTranslations` each time, even though it's doing the same work.  Let's refactor this so that `GetStringFileInfo` takes just the translation and `GetFirstStringFileInfo` takes no parameters, with both returning an object that stores the translation and has a single method `Get` or `GetString` that retrieves a string for that translation.

Reply:
I'm thinking about it. We can add GetString method to FileVersionTranslation struct. But to do it we must pass to it `data []byte` from FileVersion. So if the file has multiple translations we duplicate `data []byte` many times ...

Reply:
Let's keep them separate.  If I never call `GetTranslations` and instead manually construct a translation from a language ID and code page, you should be able to query for strings using it.  I've definitely seen code that worked that way (in fact, the first sample code that I ever saw for the version API back in the days of Windows 3.1 worked that way, if I recall correctly), so we should support that use case in Go in case someone wants it.

Reply:
We can keep them separate. But problem still exists. If I understand correctly you suggest something like this:

```
func (self *FileVersion) GetFirstStringFileInfo() (*Translation, error) {
	tr, err := self.GetTranslations()
	if err != nil || len(tr) == 0 {
		return nil, err
	}
	return &Translation{data: self.data, translation: tr[0]}, nil}
}

type Translation struct {
	data        []byte
	translation FileVersionTranslation
}

func (self *Translation) GetString(stringName FileVersionString) (string, error) {
	var offset *uint16
	var len uint32
	err := wrappers.VerQueryValue(
		&self.data[0],
		syscall.StringToUTF16Ptr(fmt.Sprintf("\\StringFileInfo\\%04x%04x\\%s", self.translation.Language, self.translation.CodePage, stringName)),
		(**byte)(unsafe.Pointer(&offset)),
		&len)
	if err != nil {
		return "", NewWindowsError("VerQueryValue", err)
	}
	if len == 0 {
		return "", nil
	}
	return LpstrToString(offset), nil
}
```
We must pass the `data []byte` to the Translation object. Sometimes `data` it can be a large array (of course we can pass pointer to array). But it will not be convenient to use. For getting string we must do something like this: 
```
	fi, err := gowin32.GetFileVersion(fileName)
	if err == nil {
		if t, err := fi.GetFirstStringFileInfo(); err == nil {
			if s, err := t.GetString(gowin32.FileVersionFileDescription); err == nil {
				// doing something with s 
			}
		}
	}

```
Are you sure about this idea ?

Reply:
No, what I'm saying is to have a new object holding the data and the translation that gets returned when you call `GetStringFileInfo` or `GetFirstStringFileInfo` and have `GetString` be a method on that object, not on `Translation.

Reply:
I think it's too late because I don't get it :) Can you show me some code ?

Reply:
Sure:
```
func (v *FileVersion) GetStringFileInfo(t Translation) *StringFileInfo { /* … */ }
func (v *FileVersion) GetFirstStringFileInfo() (*StringFileInfo, error) { /* … */ }

type StringFileInfo struct {
    translation Translation
    data        []byte
}

func (sfi *StringFileInfo) GetString(stringName FileVersionString) (string, error) { /* … */ }
```

Just curious, what time zone are you in?  It's the middle of the work day here in New York.

Reply:
Ok it's the same as I thought. Your StringFileInfo its my Translation, Translation is new type, not connected to previous types. It was a temporary name to show the code.  I am in the GMT + 1 zone. It's not late, 8 PM, but I slept too short today :)